src/Nethermind/Nethermind.Consensus.AuRa/Transactions/PermissionTxPoolOrderStrategy.cs

```diff
@@ -0,0 +1,146 @@
+﻿//  Copyright (c) 2018 Demerzel Solutions Limited
+//  This file is part of the Nethermind library.
+// 
+//  The Nethermind library is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+// 
+//  The Nethermind library is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//  GNU Lesser General Public License for more details.
+// 
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+// 
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Nethermind.Blockchain.Producers;
+using Nethermind.Consensus.AuRa.Contracts;
+using Nethermind.Consensus.Transactions;
+using Nethermind.Core;
+using Nethermind.Int256;
+
+namespace Nethermind.Consensus.AuRa.Transactions
+{
+    public class PermissionTxPoolOrderStrategy : TxPoolTxSource.ITxPoolOrderStrategy
+    {
+        private readonly IContractDataStore<Address> _sendersWhitelist;
+        private readonly IDictionaryContractDataStore<TxPriorityContract.Destination> _priorities;
+
+        public PermissionTxPoolOrderStrategy(
+            IContractDataStore<Address> sendersWhitelist, 
+            IDictionaryContractDataStore<TxPriorityContract.Destination> priorities)
+        {
+            _sendersWhitelist = sendersWhitelist ?? throw new ArgumentNullException(nameof(sendersWhitelist));
+            _priorities = priorities ?? throw new ArgumentNullException(nameof(priorities));;
+        }
+        
+        public IEnumerable<Transaction> Order(BlockHeader blockHeader, IEnumerable<Transaction> transactions)
+        {
```

Comment:
I have a feeling that all of this could have been done by a simple single comparer aand one sort
Just in the comparer compare first by IsWhitelisted, then by address, then by nonce and then by normal comparer. Code would be simpler and much much faster

Reply:
to be discussed

Reply:
This is n log(n), yours is n^2log(n), this is more optimized works same